sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -52,14 +64,122 @@
+
+/**
+ * @brief Port for the custom UDP sync protocol
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT          (16179)
+#endif
+
+/**
+ * @brief Max number of other routers on the same link
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX     (4)
+#endif
+
+/**
+ * @brief How often the number subnets should be announced by the routers
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES    (3)
+#endif
+/**
+ * @brief How long to wait for other routers annoucements before resending
+ *        or creating subnets when the retry counter is exhausted
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS    (50)
+#endif
+
+#define SERVER_THREAD_STACKSIZE                     (THREAD_STACKSIZE_DEFAULT)
+#define SERVER_MSG_QUEUE_SIZE                       (CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX)
+#define SERVER_MSG_TYPE_TIMEOUT                     (0x8fae)
 
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
 static char addr_str[IPV6_ADDR_MAX_STR_LEN];
 
+#if !IS_USED(MODULE_GNRC_IPV6_AUTO_SUBNETS_SIMPLE)
+
+/**
+ * @brief Custom UDP sync protocol
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t num_subnets;    /**< number of subnets a host wants to create */
+} auto_subnets_request_t;
+
+/* keep a copy of PIO information in memory */
+static gnrc_netif_t *_upstream;
+static ndp_opt_pi_t _pio_cache;
+
+static char auto_subnets_stack[SERVER_THREAD_STACKSIZE];
+static msg_t server_queue[SERVER_MSG_QUEUE_SIZE];
+
+/* store neighbor routers l2 address to ignore duplicate packets */
+static uint8_t l2addrs[CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX]
+                      [CONFIG_GNRC_IPV6_NIB_L2ADDR_MAX_LEN];
+
+/* PID of the event thread */
+static kernel_pid_t _server_pid;
+
+static int _send_udp(gnrc_netif_t *netif, const ipv6_addr_t *addr,
+                     uint16_t port, const void *data, size_t len)
+{
+    gnrc_pktsnip_t *payload, *udp, *ip;
+
+    /* allocate payload */
+    payload = gnrc_pktbuf_add(NULL, data, len, GNRC_NETTYPE_UNDEF);
+    if (payload == NULL) {
+        DEBUG("auto_subnets: unable to copy data to packet buffer\n");
+        return -ENOBUFS;
+    }
+
+    /* allocate UDP header, set source port := destination port */
+    udp = gnrc_udp_hdr_build(payload, port, port);
+    if (udp == NULL) {
+        DEBUG("auto_subnets: unable to allocate UDP header\n");
+        gnrc_pktbuf_release(payload);
+        return -ENOBUFS;
+    }
+
+    /* allocate IPv6 header */
+    ip = gnrc_ipv6_hdr_build(udp, NULL, addr);
+    if (ip == NULL) {
+        DEBUG("auto_subnets: unable to allocate IPv6 header\n");
+        gnrc_pktbuf_release(udp);
+        return -ENOBUFS;
+    }
+
+    /* add netif header, if interface was given */
+    if (netif != NULL) {
+        gnrc_pktsnip_t *netif_hdr = gnrc_netif_hdr_build(NULL, 0, NULL, 0);
```

Comment:
Shouldn't there be a check if this is `NULL` as well?

```suggestion
        gnrc_pktsnip_t *netif_hdr = gnrc_netif_hdr_build(NULL, 0, NULL, 0);
        
        if (netif_hdr == NULL) {
            DEBUG("auto_subnets: unable to allocate netif header\n");
            gnrc_pktbuf_release(ip);
            return -ENOBUFS;
        }
```